File: 4kyu/next_bigger_number_with_same_digits.js
function nextBigger(n) {
  let arr = String(n).split('');
  arr = arr.map(Number);
  let ans = [];
  for (let i = arr.length - 2; i >= 0; i--) {
    const a = arr[i];
    const b = arr[i + 1];
    if (a < b) {
      ans = ans.concat(arr.slice(0, i));
      const rest = arr.slice(i + 1);
      const m = Math.min(...rest.filter(t => t > a));
      ans.push(m);
      const ndx = rest.indexOf(m);
      rest.splice(ndx, 1);
      rest.unshift(arr[i]);
      rest.sort((r1, r2) => Number(r1) - Number(r2));
      ans = ans.concat(rest);
      break;
    }
  }
  if (ans.length === 0) return -1;
  return Number(ans.join(''));
}

File: 4kyu/next_bigger_number_with_same_digits.py
def biggest(n):
    t = n
    arr = []
    while t > 0:
        arr.append(t%10)
        t //= 10
        arr = sorted(arr, reverse=True)
        b = 0
    for i in arr:
        b = b * 10 + i
    return b

def next_bigger(n):
    b = biggest(n)
    if n == b:
        return -1
    n += 1
    while b != biggest(n):
        n += 1
    return n
